yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,263 @@
-from typing import Optional
-
-from docker import DockerClient
+import json
+import logging
+import selectors
+import socket
+import threading
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union, cast, TypedDict
+from weakref import WeakMethod
 
 from yellowbox.retry import RetrySpec
-from yellowbox.containers import get_ports, create_and_pull
-from yellowbox.subclasses import SingleContainerService, RunMixin
+from yellowbox.subclasses import YellowService
+from yellowbox.utils import _docker_host_name
+
+__all__ = ['LogstashService']
+_logger = logging.getLogger(__name__)
+
+_LISTEN_BACKLOG = 5  # Awaiting connections backlog count
+_STOP_TIMEOUT = 5  # Timeout for stopping the service
+
+# Sent on internal connection to signal closing of background thread
+_CLOSE_SENTINEL = b"\0"
+
+
+class LogstashService(YellowService):
+    """Implements a fake logging service that closely resembles Logstash.
+
+    Accepts external connections, with logs received in the "json_lines" format.
+
+    Attributes:
+        records: A list of all records received. You are welcome to modify,
+        clear or iterate over this during runtime. New logs will be added in the
+        order they were received. Each record is a dict - for possible keys see
+        attribute docs.
+        port: Dynamic port external service should connect to. LogstashService
+        automatically binds a free port during initialization.
+        encoding: Encoding of the json lines received. Defaults to utf-8 per
+        specification.
+        delimiter: Delimiter splitting between json objects. Defaults to b'\n' per
+        specification.
+    """
+    port: int
+    records: List[Dict[str, Any]]
```

Comment:
`TypedDict`

Reply:
You can't.

Reply:
TypedDict is covariant as I understand it, no? A dict can include additional fields without breaking usage.

Reply:
Nope

Reply:
After some checking, it appears this not the case (see https://github.com/python/mypy/issues/4617). However I still think that passing users log entries as dicts is a bad idea. Maybe an adapter object?

Reply:
At first I wanted to create LogRecord objects, but I can't recreate it well enough as I'm missing some data and logstash formats the logs differently.